File: wpimath/src/main/java/org/wpilib/math/trajectory/TrajectorySample.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.math.trajectory;

import static org.wpilib.units.Units.Seconds;

import io.avaje.jsonb.Json;
import java.util.Objects;
import org.wpilib.math.trajectory.proto.TrajectorySampleProto;
import org.wpilib.math.trajectory.struct.TrajectorySampleStruct;
import org.wpilib.units.measure.Time;
import org.wpilib.util.protobuf.ProtobufSerializable;
import org.wpilib.util.struct.StructSerializable;

/** Represents a single sample in a trajectory. */
@Json
public class TrajectorySample implements StructSerializable, ProtobufSerializable {
  /** The time of the sample relative to the trajectory start, in seconds. */
  @Json.Property("time")
  public double time;

  /** Base struct for serialization. */
  public static final TrajectorySampleStruct struct = new TrajectorySampleStruct();

  /** Base proto for serialization. */
  public static final TrajectorySampleProto proto = new TrajectorySampleProto();

  /**
   * Constructs a TrajectorySample.
   *
   * @param time The time of the sample relative to the trajectory start, in seconds.
   */
  @Json.Creator
  public TrajectorySample(double time) {
    this.time = time;
  }

  /**
   * Constructs a TrajectorySample.
   *
   * @param time The time of the sample relative to the trajectory start.
   */
  public TrajectorySample(Time time) {
    this(time.in(Seconds));
  }

  @Override
  public int hashCode() {
    return Objects.hash(time);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }

    TrajectorySample that = (TrajectorySample) o;
    return Double.compare(time, that.time) == 0;
  }
}

File: wpimath/src/main/native/cpp/trajectory/TrajectorySample.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/math/trajectory/TrajectorySample.hpp"

#include <vector>

#include "wpi/units/time.hpp"
#include "wpi/util/json.hpp"

void wpi::math::to_json(wpi::util::json& json, const TrajectorySample& sample) {
  json = wpi::util::json::object("time", sample.time.value());
}

void wpi::math::from_json(const wpi::util::json& json,
                          TrajectorySample& sample) {
  sample = TrajectorySample{wpi::units::second_t{json.at("time").get_number()}};
}

void wpi::math::to_json(wpi::util::json& json,
                        const std::vector<TrajectorySample>& samples) {
  json.set_array();
  for (const auto& sample : samples) {
    json.emplace_back(sample);
  }
}

void wpi::math::from_json(const wpi::util::json& json,
                          std::vector<TrajectorySample>& samples) {
  samples.clear();
  for (const auto& sample : json.get_array()) {
    samples.emplace_back(sample.get<TrajectorySample>());
  }
}
